### DISCUSSION TITLE
Sanitize or Count custom members

### DISCUSSION THREAD

**Discussion by hahn-kev** [VerifyTests/Verify#1149] [Q&A] [answered]
I've got a class like this which I would like to apply counting or sanitizing like is done automatically for guids and dates:
```c#
public record Commit(string Hash, string ParentHash)
```
these hashes change every test, like a Guid, I just want to verify that they are stable across many objects. For example if this is my json:
```json
[
{"hash": "abc", "parentHash": "none"},
{"hash": "xyz", "parentHash": "abc"}
]
```
then it would be sanitized into this:
```json
[
{"hash": "hash1", "parentHash": "hash2"},
{"hash": "hash3", "parentHash": "hash1"}
]
```
And I could configure this by specifying members I want to put into the same counting bucket. I'm not sure how I would do that today. If there was a way I could map a member to another value then I could implement it myself by converting the hash to something consistent, basically writing my own counter like guid or dates do.

**Comment by SimonCropp** [comment, marked as answer]
there is no built in feature to achieve this. but your could leverage MemberConverter https://github.com/VerifyTests/Verify/blob/main/docs/serializer-settings.md#converting-a-member

```
public class CounterTarget
{
    public string? Property1 { get; set; }
    public string? Property2 { get; set; }
    public string? Property3 { get; set; }
}

public class CustomCounterTests
{
    static AsyncLocal<List<string>?> local = new();

    [ModuleInitializer]
    public static void CounterInit()
    {
        VerifierSettings.OnVerify(
            before: () => local.Value = new(),
            after: () => local.Value = null
        );

        AddCountConverter<CounterTarget>( _ => _.Property1);
        AddCountConverter<CounterTarget>( _ => _.Property2);
        AddCountConverter<CounterTarget>( _ => _.Property3);
    }

    static void AddCountConverter<T>(Expression<Func<T, string?>> expression) =>
        VerifierSettings.MemberConverter<T, string?>(
            expression: expression,
            converter: ConvertMember);

    static string? ConvertMember(string? member)
    {
        if (member == null)
        {
            return null;
        }

        var values = local.Value!;
        var index = values.IndexOf(member);

        if (index == -1)
        {
            index = values.Count;
            values.Add(member);
        }

        return $"number{index + 1}";
    }

    [Fact]
    public Task CounterUsage()
    {
        var target = new CounterTarget
        {
            Property1 = "valueA",
            Property2 = "valueB",
            Property3 = "valueA",
        };
        return Verify(target);
    }
}
```

produces 

```
﻿{
  Property1: number1,
  Property2: number2,
  Property3: number1
}
```

**Reply by hahn-kev** [reply]
Thanks that worked our great! I wouldn't have thought to use the AsyncLocal.

**Reply by mattiasnordqvist** [reply]
This only works for string members right?

**Reply by hahn-kev** [reply]
the code above yes, but it could easily work for anything as long as you can easily compare the values to see that they're equal

**Reply by mattiasnordqvist** [reply]
I think the memberconverter can't convert between types, so my long-ids cant be converted to a string like "UserId_1". I solved it with some custom JsonSerializers though.
